### PR TITLE
Use BootID instead of ExternalID to check for new instance

### LINKED ISSUE #62713: Failing Tests : Timeout and UpgradeTest in gce-1.10-master-upgrade-cluster, gce-1.10-master-upgrade-cluster-new and gce-1.10-master-upgrade-cluster-parallel
### [ci-kubernetes-e2e-gce-new-master-upgrade-cluster-new](https://storage.googleapis.com/k8s-gubernator/triage/index.html?job=ci-kubernetes-e2e-gce-new-master-upgrade-cluster-new&test=Timeout)

This test is currently failing in sig-release master-upgrade suite, This could potentially become release blocking if not addressed sooner. @BenTheElder can you please triage this failure?

/kind bug
/priority failing-test
/priority important-soon
/sig testing
@kubernetes/sig-testing-bugs 

cc @jberkus @tpepper 

/assign @BenTheElder

### PR DESCRIPTION
**Pull request by krousey**
PR #60692 changed the way that ExternalID is reported on GCE. Its value
is no longer the GCE instance ID. It is the instance name. So it
cannot be used to determine VM uniqueness across time. Instead,
upgrade will check that the boot ID changed.

**What this PR does / why we need it**:
Node upgrades stall out because the external ID remains the same across upgrades now.

**Which issue(s) this PR fixes**:
Fixes #62713 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by yujuhong** [comment]
The rest looks good, but I'm not a great bash reviewer.

File: cluster/gce/upgrade.sh [hunk 1]
```diff
@@ -322,42 +319,19 @@ function do-single-node-upgrade() {
...
-    elif [[ "${new_instance_id}" == "${instance_id}" ]]; then
-      echo -n .
-    else
-      echo "Instance ${instance} recreated."
-      break
-    fi
-    sleep 1
-  done
-
   # Wait for k8s node object to reflect new instance id
```
[the remark this answers is not in the document]
Comment (krousey) [reply cluster/gce/upgrade.sh]: Done.

File: cluster/gce/upgrade.sh [hunk 2]
```diff
@@ -291,14 +291,11 @@ function upgrade-node-env() {
...
   local -r instance="$1"
-  instance_id=$(gcloud compute instances describe "${instance}" \
-    --format='get(id)' \
-    --project="${PROJECT}" \
-    --zone="${ZONE}" 2>&1) && describe_rc=$? || describe_rc=$?
-  if [[ "${describe_rc}" != 0 ]]; then
-    echo "== FAILED to describe ${instance} =="
-    echo "${instance_id}"
-    return ${describe_rc}
+  boot_id=$("${KUBE_ROOT}/cluster/kubectl.sh" get node "${instance}" --output=jsonpath='{.status.nodeInfo.bootID}' 2>&1) && kubectl_rc=$? || kubectl_rc=$?
```
Comment (mikedanese) [inline cluster/gce/upgrade.sh]: declare kubectl_rc as a local before this.

Comment (mikedanese) [reply cluster/gce/upgrade.sh]: also declare boot_id as local

File: cluster/gce/upgrade.sh [hunk 3]
```diff
@@ -322,42 +319,22 @@ function do-single-node-upgrade() {
...
-
-  # Wait for k8s node object to reflect new instance id
+  # Wait for node status to reflect a new boot ID. This guarantees us
+  # that the node status in the API is from a different boot. This
+  # does not guarantee that the status is from the upgraded node, but
+  # it is a best effort approximation.
   echo "== Waiting for new node to be added to k8s.  ==" >&2
   while true; do
-    external_id=$("${KUBE_ROOT}/cluster/kubectl.sh" get node "${instance}" --output=jsonpath='{.spec.externalID}' 2>&1) && kubectl_rc=$? || kubectl_rc=$?
+    new_boot_id=$("${KUBE_ROOT}/cluster/kubectl.sh" get node "${instance}" --output=jsonpath='{.status.nodeInfo.bootID}' 2>&1) && kubectl_rc=$? || kubectl_rc=$?
```
Comment (mikedanese) [inline cluster/gce/upgrade.sh]: declare as local

**Comment by krousey** [comment]
Properly marks all variables in that function as local.

**Pull request merged**